src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -172,13 +176,296 @@ struct TensorContract
+  SPECTRE_ALWAYS_INLINE static constexpr std::array<size_t,
+                                                    first_contracted_index::dim>
+  get_storage_indices_to_sum(
+      const std::index_sequence<Ints...>& /*dim_seq*/) noexcept {
+    constexpr std::array<size_t, num_tensor_indices>
+        lhs_contracted_multi_index =
+            ContractedLhsStructure::get_canonical_tensor_index(I);
+    constexpr std::array<size_t, first_contracted_index::dim>
+        storage_indices_to_sum = {{UncontractedLhsStructure::get_storage_index(
+            get_tensor_index_to_sum(lhs_contracted_multi_index, Ints))...}};
+
+    return storage_indices_to_sum;
+  }
+
+  /// \brief Computes a mapping between the storage indices of the contracted
+  /// LHS components and the uncontracted LHS components to sum for a
+  /// contraction
+  ///
+  /// \details
+  /// Example: If we have RHS tensor \f${R^{a}}_{abc}\f$ and we want to contract
+  /// it to the LHS tensor \f$L_{cb}\f$, then \f$L_{cb}\f$ represents the
+  /// contracted LHS, while \f${L^{a}}_{acb}\f$ represents the uncontracted
+  /// LHS. This function computes and returns a mapping between the storage
+  /// indices of (1) each component of \f$L_{cb}\f$ and (2) the corresponding
+  /// lists of components of \f${L^{a}}_{acb}\f$ to sum to compute each
+  /// component of \f$L_{cb}\f$.
+  ///
+  /// \tparam ContractedLhsNumComponents the number of components in the
+  /// contracted LHS tensor
+  /// \tparam UncontractedLhsStructure the Structure of the uncontracted LHS
+  /// tensor
+  /// \tparam ContractedLhsStructure the Structure of the contracted LHS tensor
+  /// \tparam Ints a sequence of integers from [0, `ContractedLhsNumComponents`)
+  /// \return a mapping between the storage indices of the contracted LHS
+  /// components and the uncontracted LHS components to sum for a contraction
+  template <size_t ContractedLhsNumComponents,
+            typename UncontractedLhsStructure, typename ContractedLhsStructure,
+            size_t... Ints>
+  SPECTRE_ALWAYS_INLINE static constexpr std::array<
+      std::array<size_t, first_contracted_index::dim>,
+      ContractedLhsNumComponents>
+  get_map_of_components_to_sum(
+      const std::index_sequence<Ints...>& /*index_seq*/) noexcept {
+    constexpr std::make_index_sequence<first_contracted_index::dim> dim_seq{};
+    constexpr std::array<std::array<size_t, first_contracted_index::dim>,
+                         ContractedLhsNumComponents>
+        map_of_components_to_sum = {
+            {get_storage_indices_to_sum<Ints, UncontractedLhsStructure,
+                                        ContractedLhsStructure>(dim_seq)...}};
+    return map_of_components_to_sum;
+  }
+
+  // Inserts the first contracted TensorIndex into the list of contracted LHS
+  // TensorIndexs
+  template <typename... LhsIndices>
+  using get_uncontracted_lhs_tensorindex_list_helper = tmpl::append<
+      tmpl::at_c<tmpl::split_at<tmpl::list<LhsIndices...>,
```

Comment:
Hmm I might be misunderstanding, but this is supposed to essentially append three lists: (1) list of indices before the first index, (2) a list containing just the first index, and (3) the list of indices after the first index. So in Python with lists, `get_uncontracted_lhs_tensorindex_list_helper ` is something like:

```
lhsindices_with_firstcontractedindexpos_inserted =
    LhsIndices[:FirstContractedIndexPos] +    # tmpl::at_c<tmpl::split_at<tmpl::list<LhsIndices...>, tmpl::size_t<FirstContractedIndexPos>>, 0>
    [ArgsList[FirstContractedIndexPos]]  +    # tmpl::list<tmpl::at_c<ArgsList, FirstContractedIndexPos>>
    LhsIndices[FirstContractedIndexPos:]      # tmpl::at_c<tmpl::split_at<tmpl::list<LhsIndices...>, tmpl::size_t<FirstContractedIndexPos>>, 1>
```

Based on this, I'm not sure how I could get around using `split_at` unless there is a better `brigand` function. I think I originally looked into `insert` but ran into issues - though I could revisit!

If my explanation above didn't change anything on your end, then I think maybe I'm not following the suggested change. Let me know!

Reply:
Okay, so I unfortunately don't know python well enough to understand that, but I appreciate the effort :P I see what you're doing now and it's your use of `tmpl::at_c` instead of `tmpl::front` and `tmpl::back` that have me confused. I think switching to those would make the code a lot easier to understand, at least for me. Thoughts?

Reply:
This is basically the opposite of `erase_c`, but I don't think Brigand provides a function to do that.
```c++
template <class L, unsigned int Index>
using erase_c =
    append<front<split_at<L, brigand::integral_constant<unsigned int, Index>>>,
           pop_front<back<split_at<L, brigand::integral_constant<unsigned int, Index>>>>>;
```
I think `insert` is for maps, but Cornell astro's hosting is down so I can't easily read my documentation.

Reply:
My bad about the Python! I was hoping to represent it in a more syntax-friendly way since I felt a C++ example might not be as straight-forward to read, but I'll be sure to not do that next time.

Yes to this seeming like the opposite of `tmpl::erase_c` and yes to using `tmpl::front` and `tmpl::back` probably being easier to read. I'll make this change!

Reply:
I believe that @wthrowe is right about `insert` only being for maps. Makes me wonder if an `insert`/`insert_c` for lists would be a welcome proposed addition to `brigand` or not

Reply:
I'm sure us or someone else would fine it useful. They've been a bit hit-and-miss when it comes to responding to PRs quickly (understandably for an open source project).

Reply:
`at` and `erase` are already "polymorphic" to both lists and maps (and have corresponding `_c` versions), so making `insert` work for lists would make things more consistent.